geomesa-accumulo/geomesa-accumulo-datastore/src/main/scala/org/locationtech/geomesa/accumulo/data/GeoMesaMetadata.scala

```diff
@@ -8,254 +8,280 @@
-    // also pre-fetch into the cache
-    if (!value.isEmpty) {
-      metaDataCache.synchronized { metaDataCache.put((featureName, key), Some(value)) }
+  override def read(typeName: String, key: String, cache: Boolean): Option[T] = {
+    if (!cache) {
+      metaDataCache.invalidate((typeName, key))
     }
+    metaDataCache.get((typeName, key))
   }
 
-  /**
-   * Handles writing mutations
-   *
-   * @param mutations
-   */
-  private def writeMutations(mutations: Mutation*): Unit = {
+  override def insert(typeName: String, key: String, value: T): Unit = insert(typeName, Map(key -> value))
+
+  override def insert(typeName: String, kvPairs: Map[String, T]): Unit = {
     ensureTableExists()
-    val writer = connector.createBatchWriter(catalogTable, metadataBWConfig)
-    for (mutation <- mutations) {
-      writer.addMutation(mutation)
+    val insert = new Mutation(getMetadataRowKey(typeName))
+    kvPairs.foreach { case (k,v) =>
+      val value = new Value(serializer.serialize(typeName, k, v))
+      insert.put(new Text(k), EMPTY_COLQ, value)
     }
-    writer.flush()
+    val writer = connector.createBatchWriter(catalogTable, metadataBWConfig)
+    writer.addMutation(insert)
     writer.close()
+    // also pre-fetch into the cache
+    val toCache = kvPairs.map(kv => (typeName, kv._1) -> Option(kv._2))
+    metaDataCache.putAll(toCache)
+  }
+
+  override def remove(typeName: String, key: String): Unit = {
+    if (synchronized(tableExists)) {
+      val delete = new Mutation(getMetadataRowKey(typeName))
+      delete.putDelete(new Text(key), EMPTY_COLQ)
+      val writer = connector.createBatchWriter(catalogTable, metadataBWConfig)
+      writer.addMutation(delete)
+      writer.close()
+      // also remove from the cache
+      metaDataCache.invalidate((typeName, key))
+    } else {
+      logger.warn(s"Trying to delete '$typeName:$key' from '$catalogTable' but table does not exist")
+    }
   }
 
   /**
-   * Handles deleting metadata from the catalog by using the Range obtained from the METADATA_TAG and featureName
+   * Handles deleting metadata from the catalog by using the Range obtained from the METADATA_TAG and typeName
    * and setting that as the Range to be handled and deleted by Accumulo's BatchDeleter
    *
-   * @param featureName the name of the table to query and delete from
-   * @param numThreads the number of concurrent threads to spawn for querying
+   * @param typeName the name of the table to query and delete from
    */
-  override def delete(featureName: String, numThreads: Int): Unit =
-    if (tableLock.synchronized(tableExists)) {
-      val range = new Range(getMetadataRowKey(featureName))
-      val auths = authorizationsProvider.getAuthorizations
-      val deleter = connector.createBatchDeleter(catalogTable, auths, numThreads, metadataBWConfig)
+  override def delete(typeName: String): Unit = {
+    if (synchronized(tableExists)) {
```

Comment:
the deletion should be thread safe, although obviously you wouldn't normally be calling it multiple times. The synchronization is just for the check to see if the table actually exists - we don't create the metadata table until we write something to it (so that e.g. if you fat-finger a catalog table in geoserver you don't create a random accumulo table).